### DISCUSSION TITLE
quiz object methods - get question groups

### DISCUSSION THREAD

**Discussion by vincentvmarshburn** [ucfopen/canvasapi#591] [Q&A]
I have been using canvasapi for a bit over a year now, quite successfully. Amazing work. You have my eternal gratitude.

Recently, I have been working on an application to aid with managing quiz content. While I am able to perform quite a number of operations, it has become apparent that it would be useful to be able to obtain information on all the question groups in any given quiz (classic). I know that there is a ".get_quiz_group()" method for the "Quiz" class, and I know there is a "QuizGroup" class which provides some very useful functions; however, it appears that the question group ID is necessary to begin specifying a particular question group. I have been through the documentation (and even peeked at the python source) to see if I may be missing something, but is there a method for obtaining question group IDs (or other general information on question groups) using this library? (On the other hand, I have been able to create new questions groups within a quiz, and that has been extremely helpful, thank you very much.)

Interestingly, I have gone so far as to query ChatGPT, and it unabashedly informed me that there is a ".get_question_groups()" method for a quiz object. While it sounded quite promising, and I have updated to the latest version of canvasapi, this particular method seems to elude my awareness. I know ChatGPT is not perfect, but it sounded very convincing in this matter.

Any insight would be greatly appreciated. Thank you.

**Comment by bennettscience** [comment]
Alas, ChatGPT offers seemingly correct answers that are, in fact, 100% incorrect. You're right - there is no `get_question_groups()` method in the library because there is no endpoint Canvas offers to get the question groups on a quiz. 

In Canvas, Question Groups don't exist until a student has made a submission and, at that point, they only exist on that submission. The question groups are not present at all if you try to access the quiz directly to get group IDs.

I confirmed this by making a quiz with one question and one group linked to a question bank. Calling `/api/v1/quizzes/:quiz_id/questions`, I got:

```json
  [
    {
      "id":870320,
      "quiz_id":76191,
      "quiz_group_id":null,
      "assessment_question_id":1052946,
      "position":null,
      "question_name":"Question",
      "question_type":"essay_question",
      "question_text":"This is the question",
      "points_possible":1.0,
      // ... the rest of the properties
   }
]
```

This confirmed that my question group is not included becase, at the point of creation, there are no questions in the group. Questions are selected from the attached bank at the moment a student opens the assessment. It seems, [according to this _lengthy_ answer in the community forums](https://community.canvaslms.com/t5/Canvas-Developers-Group/Listing-all-question-groups-in-a-quiz/td-p/208370), that the only way to get question group IDs via the API is to look at a student submission. I'm saying this having read James' post in the community, but I have _not_ tried it myself, so I cannot guarantee that the question group IDs you get using this method are the same for all students in the course.

To summarize James' solution (the thread is worth reading even though the code syntax got all borked when Instructure changed their forum platform), once a student has taken the assessment, you can use the Submissons endpoint to get the student submission which can then give you access to some details on the questions. The steps are:

1. Use the Quiz ID to get the submissions
2. Use a `submission_id` to get the questions submitted on that attempt.
3. Questions which are part of a question group will have a non-`null` `quiz_group_id` parameter returned.

In the library, your script would look something like this:

```python
  # imports and Canvas init
  course = canvas.get_course(1)
  submissions = course.get_quiz(2).get_submissions()
  
  for sub in submissions:
    # use the current submission ID to get the individual quiz
    quiz_submission = quiz.get_quiz_submission(sub.id)
    
    # Now you have a QuizSubmission class with a method to get all of the questions for that attempt
    questions = quiz_submission.get_submission_questions()

    # questions is a list of QuizSubmissionQuestion objects that you can finally get the group ID on
    for item in questions:
        if item.quiz_group_id is not None:
          group = quiz.get_quiz_group(item.quiz_group_id)
          # do stuff
```
I hope this helps you come up with a working solution. It definitely needs to be explored whether or not different submissions by different students show the same group IDs for the submissions. Please leave an update with what you learn!

**Reply by vincentvmarshburn** [reply]
Thank you so much for the prompt and thorough response.

I will definitely look into the submissions aspect and see if this might facilitate development of a practical technique for working with question groups. As you say, the Canvas documentation does not really indicate any direct method for identifying question group ID, which is interesting because it does include references and endpoints to such data.

In the meantime, I have resorted to adapting a method that I devised for dealing with question banks, which is another facet of the API that seems just a bit, shall we say, lacking; a somewhat odd omission if you ask me. (Otherwise, I have found the Canvas API in general and this library in particular to be quite brilliant and enormously useful.) As a kind of "workaround" for obtaining such information that does not appear to be exposed by the API directly, I managed to formulate a process that relies primarily on the python requests library and a bit of BeautifulSoup to do some minor Web scraping to get question bank info along with question group info. It is not the most elegant solution, but it seems to do what I need for the moment.

In any case, thank you again for the informative feedback. At least I can be somewhat reassured that I am not entirely daft for overlooking something that some AI insists exists but actually does not. I keep telling people to resist the hype; as impressive as current AI seems to be, it is clearly not even close to perfect. (Where did ChatGPT come up with `get_question_groups()`, I wonder? Apparently, it is sufficiently "creative" to the point of completely fabricating seemingly legitimate-sounding but fictitious information. So I suppose in that sense, it may indeed be "almost human." Fascinating.)

**Reply by bennettscience** [reply]
In fairness, I think that's what the method _would_ be called if there were an endpoint for it. This is a good example of ChatGPT returning a response based on other patterns in language, not from the _actual_ library you questioned.

Glad you have a workaround for now and that my answer was helpful.
